src/edu/washington/escience/myria/operator/agg/AggregatorFactory.java

```diff
@@ -16,18 +15,44 @@
  */
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
 @JsonSubTypes({
-  @Type(value = CountAllAggregatorFactory.class, name = "CountAll"),
-  @Type(value = SingleColumnAggregatorFactory.class, name = "SingleColumn"),
+  @Type(value = PrimitiveAggregatorFactory.class, name = "Primitive"),
   @Type(value = UserDefinedAggregatorFactory.class, name = "UserDefined")
 })
 public interface AggregatorFactory extends Serializable {
   /**
-   * Create a new aggregator for tuples of the specified schema.
+   * Generate aggregators for emitting outputs.
    *
    * @param inputSchema the schema that incoming tuples will take.
-   * @return a new aggregator for tuples of the specified schema.
-   * @throws DbException if there is an error creating the aggregator.
+   * @param offset the starting column index of aggregators made by this factory in the state hash table.
+   * @return aggregators for emitting outputs.
+   * @throws DbException if there is an error creating the aggregators
+   */
+  List<Aggregator> generateEmitAggs(Schema inputSchema, final int offset) throws DbException;
+
+  /**
+   * Generate internal aggregators, each aggregator corresponds to a column of the internal state.
+   *
+   * @param inputSchema the schema that incoming tuples will take.
+   * @param offset the starting column index of aggregators made by this factory in the state hash table.
+   * @return aggregators aggregators for internal state.
+   * @throws DbException if there is an error creating the aggregators
+   */
+  List<Aggregator> generateInternalAggs(final Schema inputSchema, final int offset)
```

Comment:
Do you think using expressions like in `Apply` would be a good idea? Actually `Aggregate` reminds me of `Apply` a lot especially after this refactoring.